### PR TITLE
feat(rcm): aggregate remote config payloads

### PR DESCRIPTION
**Pull request by avara1986**
## Checklist
New Remote Config Callback class for Remote Config payloads to group payloads by type of callback.
ASM needs to aggregate the Remote Config Agent payloads results to call `waf.update` one time for each 

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Author is aware of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer is aware of, and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

**Review by gnufede — APPROVED** [review]

**Review by christophe-papazian — APPROVED** [review]

**Review by juanjux — APPROVED** [review]

File: ddtrace/internal/remoteconfig/client.py [hunk 1]
```diff
@@ -149,6 +147,26 @@ def _load_json(data):
...
 
+class RemoteConfigCallBack(six.with_metaclass(abc.ABCMeta)):
+    @abc.abstractmethod
+    def __call__(self, metadata, config):
+        # type: (Optional[ConfigMetadata], Any) -> None
+        pass
+
+
+class RemoteConfigCallBackAfterMerge(RemoteConfigCallBack):
+    configs = {}  # type: Dict[str, Any]
```
Comment (mabdinur) [inline ddtrace/internal/remoteconfig/client.py:158]: Should the config dictionary be shared across instances?

**Review by mabdinur — COMMENTED** [review]
left a question but aside from that LGTM

File: ddtrace/internal/remoteconfig/client.py (hunk 1 above)
Comment (avara1986) [reply ddtrace/internal/remoteconfig/client.py:158]: No it doesn't. My previous approach of this OR had got this problem in tests. With this class, everything is fine :upside_down_face:

**Review by gnufede — APPROVED** [review]

**Review by juanjux — APPROVED** [review]

File: ddtrace/internal/remoteconfig/client.py [hunk 2]
```diff
@@ -366,12 +384,19 @@ def _load_new_configurations(self, applied_configs, client_configs, payload):
 
             try:
                 log.debug("Load new configuration: %s. content ", target)
-                callback(config, config_content)
+                if isinstance(callback, RemoteConfigCallBackAfterMerge):
```
Comment (P403n1x87) [inline ddtrace/internal/remoteconfig/client.py:387]: Can't this aggregation be performed in ASM? Isn't there enough information in each config to perform it outside of this class? If so, perhaps we can just change the callback mechanism to just do aggregation for each product?

Comment (avara1986) [reply ddtrace/internal/remoteconfig/client.py:387]: No, I think we can't, We can aggregate the data in ASM but we don't know in ASM side when we should report aggregate data or if it's a new payload. 

In other hand, we need to aggregate more than one product. With this approach, each team (ASM, APM, Debugger) decide what mechanism of callback is better for their needs through this new Callback class. I think it's the best approach to adapt RC callbacks for the team needs without modify the logic of other teams payloads

**Comment by avara1986** [comment]
CAVEAT: internal folder coverage isn't working correctly :face_with_spiral_eyes:

**Review by majorgreys — APPROVED** [review]

File: ddtrace/appsec/_remoteconfiguration.py [hunk 3]
```diff
@@ -134,7 +106,39 @@ def _reload_features(metadata, features):
...
+            rc_appsec_enabled = False
+        else:
+            rc_appsec_enabled = features.get("asm", {}).get("enabled")
+
+        if rc_appsec_enabled is not None:
+            from ddtrace.appsec._constants import PRODUCTS
+            from ddtrace.internal.remoteconfig import RemoteConfig
 
-    return _reload_features
+            log.debug("Reloading Appsec 1-click: %s", rc_appsec_enabled)
```
Comment (Kyle-Verhoog) [inline ddtrace/appsec/_remoteconfiguration.py:126]: Can we not use "1-click" in code and logs as I don't think it's clear what this means. Can we say remote configuration instead?

Comment (avara1986) [reply ddtrace/appsec/_remoteconfiguration.py:126]: I'll do it in a new PR :+1: thanks!

**Pull request merged**